File: algorithms/001 two sum/Solution.go
package main

import (
	"fmt"
)

func main() {
	ans := twoSum([]int{2, 7, 11, 15}, 9)
	fmt.Println(ans)
}

// Hash Table Solution Using map
func twoSum(nums []int, target int) []int {
	m := make(map[int]int)

	for i := 0; i < len(nums); i++ {
		complement := target - nums[i]

		if v, ok := m[complement]; ok {
			return []int{v, i}
		}

		m[nums[i]] = i
	}

	//return nil; // no solution (input error)
	panic("No solution")
}

File: algorithms/001 two sum/Solution.cs
class Solution {
    
    // One-pass using Dictionary
    public int[] TwoSum(int[] nums, int target) {
        
        var map = new Dictionary<int, int>();
        
        for (int i = 0; i < nums.Length; i++) {
            
            int complement = target - nums[i];
            
            if (map.ContainsKey(complement)) {
                return new int[] { map[complement], i };
            }
            
            map[nums[i]] = i;
        }
        
        // If no return yet, there is no solution.
        throw new System.ArgumentException("No solution");
    }
}